src/base/OpenMCCellAverageProblem.C

```diff
@@ -558,6 +609,77 @@ OpenMCCellAverageProblem::initialSetup()
   }
 
   setupProblem();
+
+#ifdef ENABLE_DAGMC
```

Comment:
Oh nice, I didn't know about that one. Can I use `openmc::DAGMC_ENABLED` everywhere that `ifdef ENABLE_DAGMC` is used though? For instance, in the constructor there is:

```
    _symmetry(nullptr),
#ifdef ENABLE_DAGMC
    _skinner(nullptr),
#endif
```

I feel like that wouldn't work if I swapped them.

Reply:
True ture. What we ended up doing in OpenMC was to wrap the entire C++ and header file in a `DAGMC` `#ifdef`. That kept them mostly contained the `dagmc.cpp` and `dagmc.h` files. They can still be included elsewhere without `#ifdef`'s when DAGMC is not enabled, but no functionality is added. We could do the same here in `MOABSkinner.C/H`

Reply:
Sorry I'm not sure I follow what is being suggested - could you please rephrase?

Reply:
Something like along these lines

MoabSkinner.H:

```cpp
// other includes

#ifdef DAGMC_ENABLED
#include "moab/Core.hpp"
#include "moab/Skinner.hpp"
#include "moab/GeomTopoTool.hpp"
#include "MBTagConventions.hpp"
#endif

class MoabSkinner : public GeneralUserObject {
public:
  // Constructors
  MoabSkinner(const InputParameters & parameters);

  // MOOSE methods
  static InputParameters validParams();

  virtual void execute() override;

  virtual void initialize() override;

  virtual void finalize() override;

  virtual void threadJoin(const UserObject & /* uo */) override {}


#ifdef DAGMC_ENABLED
  // MOAB methods and data members (anything from the MOAB namespace)
  cosnt std::shared_ptr<moab::Interface>& moabPtr() { return _moab; }
#endif

```

MoabSkinner.C:

```cpp
#ifndef DAGMC_ENABLED
void MoabSkinner::execute() { mooseError("DAGMC not enabled"); }

// other MOOSE methods defined here
#else
void
MoabSkinner::execute()
{
  if (_standalone)
    update();
}

// other MOAB-based method definitions here

#endif
```

Reply:
The idea is that we'd put enough functionality in the blocks where `DAGMC_ENBALED` isn't defined that we can still include it in other files without needing to keep track of where to put new `#ifdef`'s when using the `_skinner` object in the `OpenMCCellAverageProblem` source code.

Reply:
Gotcha, thanks for going into more detail. The only downside is that the test suite currently is relying on `MoabSkinner` not existing altogether when the DAGMC build is disabled, to know when to skip those tests. So if we go this route, I'll need to re-add a dummy empty object (like this `DagMCPostprocessor` I deleted earlier: https://github.com/neams-th-coe/cardinal/pull/606).

I don't have a strong opinion either way - which do you prefer?

Reply:
Can we also use `openmc::DAGMC_ENABLED` varaible in the test quite to skip those tests?

Even if not, I do prefer this approach because it contains those preprocessor statements more cleanly and makes external code easier understand, but I'm happy to defer to you on this in the end!

Reply:
No, unfortunately not. It's not C++ code, it's syntax special to CIVET. Looks like this:

```
[Tests]
  [test]
    type = Exodiff
    ...
    required_objects = 'MoabSkinner'
  []
[]
```

I like your design idea though, I do feel like it is cleaner and if we design correctly less prone to use forgetting about an ifdef. I'll update to try it out

Reply:
Actually the more I think about it, I'm not sure this is a good idea. The downsides I see:

- Need to add the dummy `DagMCPostprocessor` back in
- `MoabSkinner` will now always exist, and nothing stops the user from adding it in their input files or seeing it build with the documentation (which could be confusing)
- More code fragmentation in `MoabSkinner` to have valid "routes" for whether DAGMC is enabled/disabled.

I think I can use `openmc::DAGMC_ENABLED` in `OpenMCCellAverageProblem` in a few places to replace the ifdefs though, as well as moving more code outside of the ifdefs, so I will do that to minimize the amount of lines the ifdefs are enclosing.

Reply:
Actually, I don't think `openmc::DAGMC_ENABLED` is going to be viable on CIVET - it seems like we're configuring to just build OpenMC once with DAGMC enabled, and then separately testing with/without DAGMC in Cardinal. I'd rather not go through the steps to build OpenMC 2x if OpenMC' test suite is ultimately responsible for ensuring correct DAGMC behavior _inside of_ OpenMC.

Reply:
Thanks for trying this out, @aprilnovak! Much appreciated. Maybe we can figure out a way to do it in the future or refactor DAGMC-relevant code into an intermediate layer.

> Need to add the dummy DagMCPostprocessor back in

Agreed that the dummy `DAGMCPostProcessor` just for the sake of testing is awkward. 😖 

> MoabSkinner will now always exist, and nothing stops the user from adding it in their input files or seeing it build with the documentation (which could be confusing)

A user could have a `MoabSkinner` object in their input file with either approach and end up with a failed sim, yeah? If anything, it seems like our own implementations of those inherited `GeneralUserObject` methods and/or constructors allows us to provide a more specific error message than the one that would be provided by default for attempting to use a non-existing user object. Am I off on this?

> More code fragmentation in MoabSkinner to have valid "routes" for whether DAGMC is enabled/disabled.

Once we've provided implementations for the inherited methods of the `GeneralUserObject` for the case without DAGMC, would we really have to add any others in the future? (aside from the case where the `GeneralUserObject` interface changes).

> Further, I don't think openmc::DAGMC_ENABLED is going to be viable on CIVET - it seems like we're configuring to just build OpenMC once with DAGMC enabled, and then separately testing with/without DAGMC in Cardinal. I'd rather not go through the steps to build OpenMC 2x if OpenMC' test suite is ultimately responsible for ensuring correct DAGMC behavior inside of OpenMC.

I can sympathize with that. Having to build multiple versions of dependencies in CI isn't ideal.

Reply:
I do like the design idea, but I'd say we revisit in the future. Mostly because of the needing-to-build-OpenMC-2x on CIVET, and:

> own implementations of those inherited GeneralUserObject methods and/or constructors allows us to provide a more specific error message than the one that would be provided by default for attempting to use a non-existing user object.

True, if someone forgets to build with DAGMC and tries using an input file with `MoabSkinner`, they get an error like `object MoabSkinner not found in library!`. But currently in Cardinal basically every object can trigger an error like that in some way or another, because most objects are inside include guards for `ENABLE_NEKRS`, `ENABLE_OPENMC`, and `ENABLE_DAGMC`. It would be nice to give a more instructive error message for all of these objects, but that may take more effort to add an insertion point in the MOOSE framework wherever that error is thrown (it's not in the constructor of the object).